File: fabric/src/main/java/dev/mayaqq/estrogen/fabric/integrations/emi/recipes/EntityInteractionEmiRecipe.java
package dev.mayaqq.estrogen.fabric.integrations.emi.recipes;

import com.simibubi.create.compat.emi.recipes.CreateEmiRecipe;
import com.simibubi.create.foundation.gui.AllGuiTextures;
import dev.emi.emi.api.stack.EmiIngredient;
import dev.emi.emi.api.stack.EmiStack;
import dev.emi.emi.api.widget.SlotWidget;
import dev.emi.emi.api.widget.WidgetHolder;
import dev.mayaqq.estrogen.fabric.integrations.emi.EmiCompat;
import dev.mayaqq.estrogen.registry.recipes.EntityInteractionRecipe;
import net.minecraft.ChatFormatting;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.screens.inventory.InventoryScreen;
import net.minecraft.network.chat.Component;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.SpawnEggItem;
import net.minecraft.world.item.crafting.Ingredient;

import java.util.List;

public class EntityInteractionEmiRecipe extends CreateEmiRecipe<EntityInteractionRecipe>  {
    public EntityInteractionEmiRecipe(EntityInteractionRecipe recipe) {
        super(EmiCompat.ENTITY_INTERACTION, recipe, 177, 70, (r) -> {
        });
        this.input = List.of(EmiIngredient.of(recipe.ingredient()));
        this.output = List.of(EmiStack.of(recipe.result()));
    }

    @Override
    public void addWidgets(WidgetHolder widgets) {
        addTexture(widgets, AllGuiTextures.JEI_SHADOW, 62, 47);
        addTexture(widgets, AllGuiTextures.JEI_DOWN_ARROW, 74, 10);

        ItemStack[] matchingStacks = recipe.entity().spawnEggs();

        SlotWidget eggs = addSlot(widgets, EmiIngredient.of(Ingredient.of(matchingStacks)), 27, 38);
        if (recipe.cantBeBaby()) {
            eggs.appendTooltip(Component.translatable("recipe.entity_interaction.cant_be_baby").withStyle(ChatFormatting.GOLD));
        }

        addSlot(widgets, input.get(0), 51, 5);

        addSlot(widgets, output.get(0), 132, 38).recipeContext(this);

        widgets.addDrawable(0, 0, 0, 0, (matrices, mouseX, mouseY, delta) -> {
            int item = (int) (System.currentTimeMillis() / 1000 % eggs.getStack().getEmiStacks().size());
            EmiIngredient current = eggs.getStack().getEmiStacks().get(item);

            ItemStack stack = current.getEmiStacks().get(0).getItemStack();
            LivingEntity entity = (LivingEntity) ((SpawnEggItem) stack.getItem()).getType(stack.getOrCreateTag()).create(Minecraft.getInstance().level);
            InventoryScreen.renderEntityInInventoryFollowsMouse(matrices, 88, 55, 20, (float) -mouseX + 87, (float) -mouseY + 20, entity);
        });
    }
}

File: src/main/kotlin/dev/mayaqq/estrogen/compat/emi/recipes/EntityInteractionEmiRecipe.kt
package dev.mayaqq.estrogen.compat.emi.recipes

import dev.emi.emi.api.recipe.BasicEmiRecipe
import dev.emi.emi.api.recipe.EmiRecipeCategory
import dev.emi.emi.api.stack.EmiIngredient
import dev.emi.emi.api.stack.EmiStack
import dev.emi.emi.api.widget.SlotWidget
import dev.emi.emi.api.widget.WidgetHolder
import dev.mayaqq.estrogen.client.content.textures.RecipeTextures
import dev.mayaqq.estrogen.compat.emi.addTexture
import dev.mayaqq.estrogen.content.recipes.EntityInteractionRecipe
import dev.mayaqq.estrogen.content.recipes.getSpawnEggs
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.screens.inventory.InventoryScreen
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.SpawnEggItem
import net.minecraft.world.item.crafting.Ingredient


class EntityInteractionEmiRecipe(category: EmiRecipeCategory, val recipe: EntityInteractionRecipe) : BasicEmiRecipe(category, recipe.recipeId, 177, 70) {
    init {
        this.inputs.add(EmiIngredient.of(recipe.ingredients[0]))
        this.outputs.add(EmiStack.of(recipe.result))

    }
    override fun addWidgets(widgets: WidgetHolder) {
        widgets.addTexture(RecipeTextures.JEI_DOWN_ARROW, 62, 47)
        widgets.addTexture(RecipeTextures.JEI_SHADOW, 74, 10)

        val matchingStacks: Array<ItemStack> = recipe.entity.getSpawnEggs().toTypedArray()

        val eggs: SlotWidget = widgets.addSlot(EmiIngredient.of(Ingredient.of(*matchingStacks)), 27, 38)

        widgets.addSlot(inputs[0], 51, 5)

        widgets.addSlot(outputs[0], 132, 38).recipeContext(this)

        widgets.addDrawable(0, 0, 0, 0) { matrices, mouseX, mouseY, delta ->
            val item = (System.currentTimeMillis() / 1000 % eggs.stack.emiStacks.size).toInt()
            val current: EmiIngredient = eggs.stack.emiStacks[item]
            val stack = current.emiStacks[0].itemStack
            val entity = Minecraft.getInstance().level?.let {
                (stack.item as SpawnEggItem).getType(stack.getOrCreateTag()).create(
                    it
                )
            } as LivingEntity
            InventoryScreen.renderEntityInInventoryFollowsMouse(
                matrices,
                88,
                55,
                20,
                -mouseX.toFloat() + 87,
                -mouseY.toFloat() + 20,
                entity
            )
        }
    }
}